contrib/IECoreUSD/src/IECoreUSD/ShaderAlgo.cpp

```diff
@@ -91,6 +93,9 @@ namespace
 			}
 		}
 
+		// TODO - does this release the CompoundData?
+		r->naughty() = const_cast<IECore::CompoundData*>( IECoreScene::ShaderNetworkAlgo::collapseSplineParameters( r->naughty() ).get() );
```

Comment:
Perhaps another option is to delay the construction of the Shader till here, and pass a CompoundDataPtr to a new constructor we add for it?

Reply:
I've done this.  The one thing that spoils our nice API symmetry is that expandSplineParameters runs on ConstCompoundDataPtr ( in order to work with the parametersData coming from a const Shader being exported ), but collapseSplineParameters runs on non-const CompoundDataPtr, so it can be passed into this new constructor.  Other than that, it all looks decent.

Reply:
Taking a non-const pointer but never modifying it makes the return-original-or-modified-copy semantics harder to grasp. I wonder if we should match `expandSplineParameters()`, and then use a const cast to pass to the shader constructor. At that point, we do know enough to be sure this is safe.